### DISCUSSION TITLE
How to set compression on backup

### DISCUSSION THREAD

**Discussion by KamikazeePL** [rustic-rs/rustic#724] [Q&A] [answered]
In **restic** I can use `--compression` option to set compression level when I'm making backup files.

How to set compression level in rustic when I make backup?

**Comment by aawsome** [comment, marked as answer]
In rustic, compression level is a repository option, i.e. it is saved in the repository config and needs to be set only once (or when you want to change it).

You can already set it when initializing a repository and change it using the command `rustic config --set-compression`.

**Reply by Schweber** [reply]
I can set
``` 
[repository]
set-compression = 9
```
in the config and rustic doesn't throw an error. Is this setting respected by rustic or silently disregarded?

As it's possible to set
```
[backup]
init = true
```
i would like to set the compression-level in the config as well.

**Reply by aawsome** [reply]
@Schweber this doesn't work. Use `rustic config --set-compression`.

**Reply by Schweber** [reply]
Ok, thanks

**Comment by aawsome** [comment]
Also note, that there is no `max` compression level, but you can use the compression levels used by `zstd`, see `rustic config --help` for more information.

**Reply by zxcat** [reply]
is it possible to view repo config?
* `rustic config --help` only shows `--set-*` flags, not `--get-*`
* and `restic repoinfo` doesn't show any config info

**Reply by aawsome** [reply]
Yes, you can use `rustic cat config`. This will, however, only show the raw json of the config file and not the default values used if an item is missing...
If anyone is up with a PR to also show the values, we'll happily merge that!

**Comment by LeoniePhiline** [comment]
For people looking to pick the right compression level:

At zstd compression levels > 3, throughput becomes too slow to be worth the little gain in compression ratio in most cases.

[Borg uses zstd compression level 3 by default](https://borgbackup.readthedocs.io/en/stable/usage/help.html?highlight=zstd#borg-help-compression).

**Reply by simonsan** [reply]
@aawsome that's something we should adapt as well, I guess. Or at least have something in the docs about it

**Reply by aawsome** [reply]
@LeoniePhiline Thanks for the information!
I must say I did not make any measurements with different compression levels. I think it also depends about what you are trying to optimize for. 

So if you optimize for backup speed, with a remote repository the backup may depend a lot on the connection to your storage backend. If you have a slow connection, but a fast CPU, higher compression rates can therefore lead to higher backup speed. For fast connections or backups to local storage, you may be better of with a low compression.

Besides this, people might optimize for other things like total cost where a higher compression helps, of course.

**Reply by LeoniePhiline** [reply]
Yes. I don't have links handy right now, but you can find benchmark charts online which show only a slight increase of compression ratio paired with a much larger increase in processing time for levels above 3.
I would expect people who want a small repo, but not a really really really super small repo to go with about lvl 3. Since disk space is cheap nowadays, and speed makes a difference, many might like to lower the level for worse compression but faster processing.

I had just wanted to mention this because zstd goes up to lvl 22, but most of these levels are not useful for everyday cases - which is not intuitive at all. Without charts (plotting compression-ratio-to-processing-time per each available level), the levels mean nothing to users.

**Reply by KamikazeePL** [reply]
https://peazip.github.io/fast-compression-benchmark-brotli-zstandard.html

**Reply by LeoniePhiline** [reply]
> https://peazip.github.io/fast-compression-benchmark-brotli-zstandard.html

Note that these zstd compression levels are prefixed `-` as if they were arguments, not as if they were negative numbers. The levels discussed here are all `> 0`, even though they show up in the table (see below) looking like negative numbers.

![grafik](https://github.com/rustic-rs/rustic/assets/22329650/cc77b2b0-16f7-40dd-bf03-86899861cd04)

The fastest (`< 0` / negative level number) zstd compression levels, which are faster and provide lower compression ratio than level `1`, were introduced later.

**Comment by aawsome** [comment]
> @aawsome that's something we should adapt as well, I guess. Or at least have something in the docs about it

Actually the default compression level (i.e. if not explicitely set by the user) is taken from the `zstd` crate. Had a quick look: 
https://docs.rs/zstd-sys/latest/zstd_sys/constant.ZSTD_CLEVEL_DEFAULT.html

The zstd crate (and therefore also rustic) uses 3 as default level.
